FlowCrypt/Functionality/DataManager/Encrypted Storage/KeyChainService.swift

```diff
@@ -19,67 +19,85 @@ actor KeyChainService {
     private let logger = Logger.nested("KeyChain")
     private let keyByteLen = 64
 
-    /// this dynamic keychainIndex ensures that we use a different keychain index
-    ///   after deleting the app, because keychain entries survive app uninstall
-    @MainActor private func getKeychainIndex() throws -> String {
-        guard UIApplication.shared.isProtectedDataAvailable else {
-            // when not available, UserDefaults is empty and a new index would be wrongly generated
-            // which would cause new database encryption key to get generated
-            // and that would corrupt existing storage
-            // https://github.com/FlowCrypt/flowcrypt-ios/issues/1373
-            throw AppErr.general("KeyChainService: protected data is not available")
-        }
-        let dynamicPartIndex = "indexSecureKeychainPrefix"
-        if let storedDynamicPart = UserDefaults.standard.string(forKey: dynamicPartIndex) {
-            return constructKeychainIndex(dynamicPart: storedDynamicPart)
-        }
-        let newDynamicPart = try newRandomString()
-        UserDefaults.standard.set(newDynamicPart, forKey: dynamicPartIndex)
-        return constructKeychainIndex(dynamicPart: newDynamicPart)
+    private enum Constants {
+        static let legacyKeychainIndexPrefixInUserDefaults = "indexSecureKeychainPrefix"
+        static let keychainPropertyKey = "app-storage-encryption-key"
     }
 
-    @MainActor private func newRandomString() throws -> String {
-        logger.logInfo("newRandomString - generating new KeyChain index")
-        guard let randomBytes = CoreHost().getSecureRandomByteNumberArray(12) else {
-            throw AppErr.general("KeyChainService.newRandomString - randomBytes are nil")
+    @MainActor var storageEncryptionKey: Data {
+        get throws {
+            if let key = try fetchEncryptionKey() {
+                return key
+            }
+
+            // TODO: Should be removed after all users migrated to new keychain
+            if let legacyKey = try fetchLegacyEncryptionKey() {
+                try saveStorageEncryptionKey(data: legacyKey)
+                return legacyKey
+            }
+
+            return try generateStorageEncryptionKey()
         }
-        return Data(randomBytes)
-            .base64EncodedString()
-            .replacingOccurrences(of: "[^A-Za-z0-9]+", with: "", options: [.regularExpression])
     }
 
-    @MainActor private func constructKeychainIndex(dynamicPart: String) -> String {
-        return dynamicPart + "-indexStorageEncryptionKey"
+    @MainActor private func fetchLegacyEncryptionKey() throws -> Data? {
+        let userDefaults = UserDefaults.standard
+        let prefixKey = Constants.legacyKeychainIndexPrefixInUserDefaults
+
+        guard let storedDynamicPart = userDefaults.string(forKey: prefixKey) else { return nil }
+
+        userDefaults.removeObject(forKey: prefixKey)
```

Comment:
I think this is a bit dangerous here - there is some chance that the above code successfully gets the prefix, but the code below fails before the actual key it's migrated to new place in key chain. That could cause trouble or make it harder to debug problems later.

Therefore I propose we only delete this from user defaults after we've successfully stored the key at new location.

Reply:
Fixed ✅